cpp/src/parquet/column_writer.cc

```diff
@@ -323,6 +323,19 @@ class SerializedPageWriter : public PageWriter {
 
     PARQUET_THROW_NOT_OK(sink_->Write(output_data_buffer, output_data_len));
 
+    // Add the dictionary page to the offsets
+    if (offset_index_builder_ != nullptr) {
+      const int64_t compressed_size = output_data_len + header_size;
+      if (compressed_size > std::numeric_limits<int32_t>::max()) {
+        throw ParquetException("Compressed dictionary page size overflows to INT32_MAX.");
+      }
+
+      /// start_pos is a relative offset in the buffered mode. It should be
+      /// adjusted via OffsetIndexBuilder::Finish() after BufferedPageWriter
+      /// has flushed all data pages.
+      offset_index_builder_->AddPage(start_pos, static_cast<int32_t>(compressed_size), -1);
+    }
```

Comment:
I think this may be a bug or oversight in the existing implementation? When decoding data for a rowgroup, if a dictionary is enabled you need to read the dictionary page. Currently this dictionary page is not written to the OffsetIndex section and, without it, there is no reliable way to know where the dictionary page can be found (due to variable sizes). Therefore, and since it is a type of data page, I believe it should be written to the OffsetIndex collection. Anyway, the current implementation needs it to decode dictionary data.

Reply:
I'm don't think this is an oversight.  Dictionary page offset is stored in [column metadata](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L805) and the [first data page offset](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L799) is also stored, the length should be the difference of these two.  CC @wgtmac to confirm.

Reply:
Yes, the difference between offsets to dict page and 1st data page can be used as the upper-bound (exact in most cases) size of the dict page.

Reply:
@emkornfield @wgtmac 
So, what you are saying is correct, but it seems I have not clearly explained my reasoning here.
The idea behind [PageIndex](https://github.com/apache/parquet-format/blob/master/PageIndex.md) is to make "point lookups I/O efficient". So, you can look at a ColumnIndex to find a range of values for a particular column, then use the OffsetIndex to efficiently load that data. For directly encoded data, the OffsetIndex is great, you can directly load and examine column data. For dictionary encoded data there is a problem. You can't read the data without having the dictionary. So, I believe that the OffsetIndex should contain the dictionary page for two reasons:
1. Without the dictionary page you cannot decode the data in subsequent data pages. Instead, you now have to go retrieve the information from the metadata. This violates the principle that the PageIndex should be a way to directly access data.
2. The dictionary page, is in fact, a type of data page. So including it should be backwards compatible since readers should recognize its type in the list of pages provided by an OffsetIndex.

Sadly, the spec seems to be silent on what should happen in the case of dictionary encodings. It seems that most writers do not provide this information. I wonder if @mkornacker or @lekv considered dictionary encodings or would be willing to comment?

Reply:
At any rate, in this PR I wanted to be able to decode data directly without having to read the full metadata header. For dictionary pages I was unable to guess the location just from the data page (because dictionary pages can have varying sizes). So, I added this information to the PageOffset since I felt it made sense to have it there, but this may not be within the PageIndex spec.

Reply:
We should clarify the spec, but I believe dictionary pages were intended to be retrieved via column metadata (which would be in the footer).  I think it would have been mentioned if it is, and I think the spec always assumes the footer is parsed, and for page indexes there is a copy of the metadata there.  The OffsetIndex is meant to directly parallel the ColumnIndex which only keep stats for data pages.

> The dictionary page, is in fact, a type of data page. So including it should be backwards compatible since readers should recognize its type in the list of pages provided by an OffsetIndex.

In the sense that everything is "data" that is correct.  But in parquet dictionaries are [modeled](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L603) as a separate concept.   The actual type of page is not stored in either index, it is therefore not really possible to make this a compatible change because old readers would likely error out here.  [Parquet-mr does not touch the index when writing a dictionary page](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L1042) either it does for [data pages](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L781).

Reply:
@emkornfield To be specific I do actually mean dictionary page in a much more specific way. As seen in the
[spec:](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L559)
```
enum PageType {
  DATA_PAGE = 0;
  INDEX_PAGE = 1;
  DICTIONARY_PAGE = 2;
  DATA_PAGE_V2 = 3;
}
```

And this is reflected in the [PageHeader](https://github.com/apache/parquet-format/blob/eb4b31c1d64a01088d02a2f9aefc6c17c54cc6fc/src/main/thrift/parquet.thrift#L693)
```
struct PageHeader {
  /** the type of the page: indicates which of the *_header fields is set **/
  1: required PageType type

  /** Uncompressed page size in bytes (not including this header) **/
  2: required i32 uncompressed_page_size

  /** Compressed (and potentially encrypted) page size in bytes, not including this header **/
  3: required i32 compressed_page_size

  /** The 32-bit CRC checksum for the page, to be be calculated as follows:
   ...
   */
  4: optional i32 crc

  // Headers for page specific data.  One only will be set.
  5: optional DataPageHeader data_page_header;
  6: optional IndexPageHeader index_page_header;
  7: optional DictionaryPageHeader dictionary_page_header;
  8: optional DataPageHeaderV2 data_page_header_v2;
}
```
So that the dictionary page is just one kind of data page. This is reflected in the reader class as in column_page.h:150
```
class DictionaryPage : public Page {
...
}
```


So, when dictionary encoded data is written to the file, first a dictionary page is written, then data pages are written. The parquet reader enforces and leverages this fact. It actually reads from a single collection of pages where it requires that the first page be a dictionary page for dictionary encoded pages. So, including the encoding page in the OffsetIndex seems pretty natural to me.

Reply:
I think we are in agreement on "a dictionary page is a type of page".   The original statement I was responding to was:

> The dictionary page, is in fact, a type of data page.

Which as highlighted above data pages have distinct enums and header metadata. It could be PageIndex was originally intended for all page types, in actual implementations and given existing metadata the index captures it is only for data pages.  Any changes to these semantics pose challenges for compatibility.